File: leetcode2/2medium/최원준/Q890.java
package Leetcode.최원준;

/*
1. 아이디어 :


2. 시간복잡도 :
O( )

3. 자료구조/알고리즘 :

 */

import java.util.*;

public class Q890 {
    class Solution {
        String pattern;
        int n;
        Set<Integer>[] pat;
        public boolean doesMatch(String word) {
            Map<Character, Character> map = new HashMap<>();
            Set<Character> visited = new HashSet<>();
            StringBuilder sb = new StringBuilder();
            for (int i=0; i<word.length(); i++) {
                char c = word.charAt(i);
                char oriChar = pattern.charAt(i);
                if (map.containsKey(c)) {
                    char mapChar = map.get(c);
                    sb.append(mapChar);
                } else {
                    if (visited.contains(oriChar)) return false;
                    visited.add(oriChar);
                    map.put(c, oriChar);
                    sb.append(oriChar);
                }
            }
            return pattern.equals(sb.toString());
        }

        public List<String> findAndReplacePattern(String[] words, String pattern) {
            this.pattern = pattern;
            n = pattern.length();
            pat = new HashSet[26];
            for (int i=0; i<26; i++) pat[i] = new HashSet<>();
            for (int i=0; i<n; i++) pat[pattern.charAt(i)-'a'].add(i);

            List<String> ans = new ArrayList<>();
            for (String word: words) {
                if (doesMatch(word)) ans.add(word);
            }
            return ans;
        }
    }
}

File: leetcode2/2medium/최은서/Q890.js
/**
 * @param {string[]} words
 * @param {string} pattern
 * @return {string[]}
 */
var findAndReplacePattern = function (words, pattern) {
  const res = [];
  for (const w of words) {
    const pm = new Map(); //pattern -> word
    const wm = new Map(); //word -> pattern
    let flag = false;
    for (let i = 0; i < w.length; i++) {
      //각 문자 하나씩 비교
      if (
        (pm.has(pattern[i]) && pm.get(pattern[i]) !== w[i]) || // pattern -> word 혹은 word -> pattern에서 중복 발생
        (wm.has(w[i]) && wm.get(w[i]) !== pattern[i])
      ) {
        flag = true;
      }
      if (flag) break;
      pm.set(pattern[i], w[i]);
      wm.set(w[i], pattern[i]);
    }
    if (!flag) res.push(w);
  }
  return res;
};
